### PR TITLE
Add different output modes: Caps, Title, Lowercase, Snake, Camel

### PR DESCRIPTION
**Pull request by morinted**
This pull adds several different cases. Namely:

- CAPITAL CASE
- Title Case
- lower case
- snake_case
- CamelCase

They are controlled by "MODE" commands, similar to "PLOVER" commands. They can be toggled with CAPS, TITLE, LOWER, SNAKE, and CAMEL. They can all be turned off with CLEAR.

CAPS, Title case, and lower case are mutually exclusive, so turning on one will turn off the others. CamelCase is mutually exclusive to all other cases. Snake case is only exclusive with camel case.

This means CAPS, title, and lower case can be used with snake case. This functionality is useful for programmers and regular stenographers alike. The cases are useful for programmers while coding, such as in variable names. For regular stenographers, CAPS can represent yelling, or scene cues, for example. Title case is useful for journalism.

I've tried to keep some sanity while coding these. The initial implementation used global variables, but this wasn't testable. Now, a dict of the cases has been added as a variable to the Action class. Since the cases are persistent, the case is passed along on each stroke. The MODE command directly modifies the mode of the action.

Some sample strokes for testing:

    "KA*PS": "{MODE:CAPS}",
    "K-BGS": "{MODE:CAMEL}",
    "TAO*EULT": "{MODE:TITLE}",
    "STPHA*EUBG": "{MODE:SNAKE}",
    "HRO*ER": "{MODE:LOWER}",
    "KHRAO*ER": "{MODE:CLEAR}",

Hope the code is up to standard -- if not, I'd be happy to make it so.

Ted

### CODE REVIEW

**Comment by hhirsch** [comment]
Why isn't this merged?

**Comment by morinted** [comment]
Two things other than the quality of code itself that I can think of: title
case doesn't work properly with multi-part words, especially because python
applies it to words like "That'S". Maybe we could make our own local title
case function to be more useful for most Plover users. Second thing: maybe
a counter could be implemented. Right now we have the ability to change
modes, it might be cool to add a counter. Omitted or 0 means "always
on/infinite", and 1 would mean "next one translation" and 10 would mean
"next 10 translations". That way, you could say "caps lock mode for 1 word"
and then do it and not have to remember to clear it. That would speed up
certain things. Or you could have a stroke for snake case for only 2
strokes if you wanted to use it to append strokes. (also, I specify
translation because of multi-word entries. I think that counting words and
trying to define what's a word and what isn't is no fair to the program or
its users. One translation is clear.)

In addition to those, Hesky would have to review my code, and I imagine
that it's not the ideal implementation, as I "wrote first, designed later"
with this one. This is the kind of thing that might warrant a rewrite if we
can thing of a better design. That being said, Hesky, if you can think of a
better design, I'd be willing to rewrite as I think this feature is
critical for Plover, and I use it now as a regular feature on my end of the
woods.

Ted

**Comment by balshetzer** [comment]
The main reason why a pull requests isn't merged, unless I've given another reason, is just that I haven't had the time for a full review and testing.

**Comment by hhirsch** [comment]
Life is short. I pull it now and we can fix it later.

File: plover/formatting.py [hunk 1]
```diff
@@ -733,37 +795,131 @@ def _atom_to_action_spaces_after(atom, last_action):
...
+
+
+def _apply_case(text, case, begin):
+    action = _Action()
+    if case == action.cases['LOWER']:
+        text = text.lower()
+    elif case == action.cases['UPPER']:
+        text = text.upper()
+    elif case == action.cases['TITLE']:
+        text = text.title()
```
Comment (morinted) [inline plover/formatting.py]: .title() is not ideal because it title cases "isn't" as "Isn'T".

**Comment by morinted** [comment]
@mighele @chajadan same for this one, time to review if you can -- I fixed the issues that I had with this. Test it out :smile_cat: Title case now works quite well!

**Comment by morinted** [comment]
I really hate title case, I must say. Now when fingerspelling, subsequent capitals get lowercased. For example, if I fingerspell "CAT" the output will invariably be "Cat". While it is "true" title case, I don't like this behavior. It would turn iPhone into Iphone or iphone (depending on if the user has set their stroke with {>})

I'll be investigating this tonight.

**Comment by morinted** [comment]
The way I addressed this was to actually ignore glued/appended strokes with Title Casing. The Way That It Works Is Basically, As I Type Output Is Titled. However, Using A Manual Space Or A TK-LS Or Any Glue Command Will Set Plover To {^} Mode. This Is Interesting Because Now if I Wanted to Lowercase Certain Words in a Sentence All I Need to do is Space ("S-P") and My Output Is Normal.

Furthermore, it's cool because now the user can fingerspell without the case mode getting in the way, just in case they need to type something styled like ShoeBOX or iPhone in a title.

**Comment by morinted** [comment]
Hi @benoit-pierre, if you could, I'd appreciate if you could check this PR as well. I'm hoping to merge it, still.

**Comment by benoit-pierre** [comment]
One test is failing:

```python

self = <plover.test_formatting.FormatterTestCase testMethod=test_apply_case>

    def test_apply_case(self):
        an_action = action()
        # case, text, begin
        test = ' some test '
        test2 = 'test Me'
        test3 = ' SOME TEST '
        cases = [
            # UNDEFINED
            ((test, '', False), (test)),
            ((test, 'TEST', False), (test)),
            # TITLE
            ((test, an_action.cases['TITLE'], False), (' Some Test ')),
            ((test, an_action.cases['TITLE'], True), (' Some Test ')),
            # TITLE responds to 'lower'
            ((test2, an_action.cases['TITLE'], True), ('test Me')),
            # LOWER
            ((test, an_action.cases['LOWER'], False), (' some test ')),
            ((test3, an_action.cases['LOWER'], False), (' some test ')),
            ((test2, an_action.cases['LOWER'], True), ('test me')),
            # UPPER
            ((test.upper(), an_action.cases['UPPER'], False), (' SOME TEST ')),
            ((test3, an_action.cases['UPPER'], False), (' SOME TEST ')),
            ((test2, an_action.cases['UPPER'], True), ('TEST ME')),
        ]

>       self.check_arglist(formatting._apply_case, cases)

plover/test_formatting.py:858:

plover/test_formatting.py:53: in check_arglist
    self.assertEqual(actual, expected)
E   AssertionError: ' some test ' != ' Some Test '

```

**Comment by morinted** [comment]
Ah, yeah, I redefined the behavior of Title Case to be less intrusive. Updated the test to reflect the behavior.

File: plover/formatting.py [hunk 2]
```diff
@@ -415,6 +454,7 @@ def _atom_to_action_spaces_before(atom, last_action):
     last_upper = last_action.upper
     last_upper_carry = last_action.upper_carry
     last_orthography = last_action.orthography
+    begin = False  # for meta attach
```
Comment (morinted) [inline plover/formatting.py]: Don't think I'm using this anymore

File: plover/formatting.py [hunk 3]
```diff
@@ -347,6 +372,17 @@ def _translation_to_actions(translation, last_action, spaces_after):
 META_ATTACH_FLAG = '^'
 META_KEY_COMBINATION = '#'
 META_COMMAND = 'PLOVER:'
+META_MODE = 'MODE:'
+MODE_CAPS = 'CAPS'
```
Comment (morinted) [inline plover/formatting.py]: I'm thinking these should be put into a tuple, MODES

File: plover/formatting.py [hunk 4]
```diff
@@ -187,8 +196,13 @@ class _Action(object):
     context to render future translations.
 
     """
-    def __init__(self, attach=False, glue=False, word='', capitalize=False, 
-                 lower=False, upper=False, upper_carry=False, orthography=True, text='', replace='', combo='',
+
+    cases = {'NONE': 0, 'UPPER': 1, 'LOWER': 2, 'TITLE': 3}
```
Comment (morinted) [inline plover/formatting.py]: Rename to CASES?

**Comment by morinted** [comment]
This wasn't working with "after" spaces. I fixed that. Also made the changes outlined above, except for my MODES set. Discovered bug: if a stroke is undefined, the mode gets unset. This needs to be fixed.

**Comment by morinted** [comment]
Thinking about it a little more, I don't mind losing the case mode on misstroke. @benoit-pierre, I'm going to merge this. Any last words before it's in?

File: plover/formatting.py [hunk 5]
```diff
@@ -187,8 +196,13 @@ class _Action(object):
     context to render future translations.
 
     """
-    def __init__(self, attach=False, glue=False, word='', capitalize=False, 
-                 lower=False, upper=False, upper_carry=False, orthography=True, text='', replace='', combo='',
+
+    CASES = {'NONE': 0, 'UPPER': 1, 'LOWER': 2, 'TITLE': 3}
```
Comment (benoit-pierre) [inline plover/formatting.py]: IMHO this would be better:
```python
CASE_NONE, CASE_UPPER, CASE_LOWER, CASE_TITLE = range(4)
```

File: plover/formatting.py [hunk 6]
```diff
@@ -187,8 +196,13 @@ class _Action(object):
...
+
+    def __init__(self, attach=False, glue=False, word='', capitalize=False,
+                 lower=False, orthography=True, space_char=' ',
+                 upper=False, upper_carry=False,
+                 case=0, text='', replace='', combo='',
```
Comment (benoit-pierre) [inline plover/formatting.py]: Should be:
```python
                 case=CASES['NONE'], text='', replace='', combo='',
```

File: plover/formatting.py [hunk 7]
```diff
@@ -347,6 +372,17 @@ def _translation_to_actions(translation, last_action, spaces_after):
...
+MODE_CAPS = 'CAPS'
+MODE_TITLE = 'TITLE'
+MODE_LOWER = 'LOWER'
+MODE_SNAKE = 'SNAKE'
+MODE_SET_SPACE = 'SET_SPACE:'
+MODE_CLEAR_SPACE = 'CLEAR_SPACE'
+MODE_CLEAR_CASE = 'CLEAR_CASE'
+MODE_CAMEL = 'CAMEL'
+MODE_CLEAR = 'CLEAR'
+
```
Comment (benoit-pierre) [inline plover/formatting.py]: I think `RESET` instead of `CLEAR` might be clearer.

File: plover/formatting.py [hunk 8]
```diff
@@ -730,44 +794,147 @@ def _atom_to_action_spaces_after(atom, last_action):
...
+    command should be:
+        CAPS, LOWER, TITLE, CAMEL, SNAKE, CLEAR_SPACE,
+            CLEAR_CASE, SET_SPACE or CLEAR
+
+        CAPS: Uppercase
+        LOWER: Lowercase
+        TITLE: Titlecase
+        CAMEL: Titlecase, no space
+        SNAKE: Underscore space
+        CLEAR_SPACE: no space
```
Comment (benoit-pierre) [inline plover/formatting.py]: Or rather: `CLEAR_SPACE: reset space character to default (' ')`, no?

File: plover/formatting.py [hunk 9]
```diff
@@ -730,44 +794,147 @@ def _atom_to_action_spaces_after(atom, last_action):
...
+    else:
+        return text
+
+
 def _get_meta(atom):
     """Return the meta command, if any, without surrounding meta markups."""
     if (atom is not None and
         atom.startswith(META_START) and
-        atom.endswith(META_END)):
+            atom.endswith(META_END)):
```
Comment (benoit-pierre) [inline plover/formatting.py]: Why the indentation change?

File: plover/formatting.py [hunk 10]
```diff
@@ -730,44 +794,147 @@ def _atom_to_action_spaces_after(atom, last_action):
...
     return atom.replace(META_ESC_START, META_START).replace(META_ESC_END,
                                                             META_END)
 
+
 def _get_engine_command(atom):
     """Return the steno engine command, if any, represented by the atom."""
     if (atom and
         atom.startswith(META_START + META_COMMAND) and
-        atom.endswith(META_END)):
+            atom.endswith(META_END)):
```
Comment (benoit-pierre) [inline plover/formatting.py]: Ditto.

File: plover/test_formatting.py [hunk 11]
```diff
@@ -721,9 +757,116 @@ def test_atom_to_action_spaces_after(self):
...
+    def test_apply_space_char(self):
+        # (text, space_char)
+        test = ' some test '
+        test2 = 'don\'t'
+        cases = [
+            ((test, '_'), ('_some_text_')),
+            ((test, ''), ('sometext')),
+            ((test2, '_'), (test2)),
+            ((test2, ''), (test2)),
+        ]
```
Comment (benoit-pierre) [inline plover/test_formatting.py]: No checks?

**Comment by morinted** [comment]
Thanks for the feedback! I'll get to this soon.

File: plover/formatting.py (hunk 9 above)
Comment (morinted) [inline plover/formatting.py]: PEP8 E125,

Wants a difference in the indentation from the next line to show separation:

1. Valid:

```python
if (thing and
    and thing
        and thing):
    do_thing()
```

2. Invalid:

```python
if (thing and
    and thing
    and thing):
    do_thing()
```

3. Valid:

```python
if (thing and
        and thing
        and thing):
    do_thing()
```

I could switch to (3).

Comment (benoit-pierre) [inline plover/formatting.py]: Meh, 1 looks ugly, sadly 2 is invalid, 3 is better then... It would have been better if those formatting and whitespace changes had been done in a separate commit: make using `git blame` easier.

Comment (morinted) [inline plover/formatting.py]: You're right. I was a bad person. :cry:

**Comment by morinted** [comment]
All right, I fixed the loss of case when you use an undefined stroke, and I updated according to your review.

**Comment by morinted** [comment]
If all is good, I'll squash and merge, just give me the go ahead.

**Comment by benoit-pierre** [comment]
:+1:

**Comment by arxanas** [comment]
For anyone reading this after the fact, `MODE:CLEAR` is now called `MODE:RESET`. The full list of mode strings can be consulted on [L375-384](https://github.com/openstenoproject/plover/pull/215/files#diff-7c608006e88cb876893096ed894f2577R375). Not sure if the modes are documented anywhere else.

**Comment by morinted** [comment]
Only on the release notes, I plan on making a dictionary format doc which
will cover case modes, key combos, and more.

**Comment by davidkitfriedman** [comment]
So with no mode setting doing TUPBT/S correctly yields: 

"opportunities"

But with the `MODE:CAPS` setting doing TUPBT/S yields:

"OPPORTUNITYIES"

I haven't looked further, but other words that end in "Y" work fine in both modes such as "city", and "sky".
